Add executable gossip validation functions for fulu (#5246)
ethereum/consensus-specs: specs/fulu/partial-columns/p2p-interface.md

    has_header = len(sidecar.header) == 1
    num_cells_present = sum(1 for b in sidecar.cells_present_bitmap if b)
    has_cells = num_cells_present > 0

    # [REJECT] A header and/or cells are present in the message
    if not (has_header or has_cells):
        raise GossipReject("partial message is semantically empty")

    # [REJECT] The cell count equals the number of bits set in cells_present_bitmap
    if len(sidecar.partial_column) != num_cells_present:
        raise GossipReject("number of cells does not match number of set bits")

    # [REJECT] The proof count equals the number of bits set in cells_present_bitmap
    if len(sidecar.kzg_proofs) != num_cells_present:
        raise GossipReject("number of proofs does not match number of set bits")

    if has_header:
        header = sidecar.header[0]
        block_header = header.signed_block_header.message

        # [REJECT] The received header MUST equal any previously validated header for this block
        prior_header = seen.partial_data_column_headers.get(block_root)
        if prior_header is not None and prior_header != header:
            raise GossipReject("header differs from previously validated header")

>         # [REJECT] The signed_block_header hash matches the partial message's group id

Reviewer: I'm not familiar with partial columns, but I guess the group ID of partial message is defined as block_root?

Also, `tests/formats/networking/gossip_validation.md` says block_root as a group "root", which is confusing to me.

Author: 🤷 I'm still a bit confused by group IDs. But yes, I think so. For reference:

* https://github.com/ethereum/consensus-specs/pull/4558#discussion_r2926849578

Reviewer: It doesn't scroll so can't see the comment but not a big of deal to me.

        if hash_tree_root(block_header) != block_root:
            raise GossipReject("header's block root does not match partial message group id")

        # [REJECT] The header's kzg_commitments list is non-empty
        if len(header.kzg_commitments) == 0:

>             raise GossipReject("header kzg_commitments is empty")

Reviewer: ```suggestion
            raise GossipReject("header's kzg_commitments is empty")
```

            raise GossipReject("invalid proposer signature on header")

        # [IGNORE] The header's block's parent has been seen
        # (MAY be queued for processing once the parent block is retrieved)
        if block_header.parent_root not in store.blocks:
            raise GossipIgnore("header's parent has not been seen")

        # [REJECT] The header's block's parent passes validation
        if block_header.parent_root not in store.block_states:
            raise GossipReject("header's parent failed validation")

        # [REJECT] The header is from a higher slot than the header's block's parent
        if block_header.slot <= store.blocks[block_header.parent_root].slot:
            raise GossipReject("header is not from a higher slot than its parent")

        # [REJECT] The current finalized_checkpoint is an ancestor of the header's block
        checkpoint_block = get_checkpoint_block(
            store, block_header.parent_root, store.finalized_checkpoint.epoch
        )
        if checkpoint_block != store.finalized_checkpoint.root:
            raise GossipReject("finalized checkpoint is not an ancestor of header's block")

>         # [REJECT] The header's kzg_commitments inclusion proof is valid

Reviewer: ```suggestion
        # [REJECT] The header's kzg_commitments inclusion proof is valid as verified by verify_partial_data_column_header_inclusion_proof
```

Author: I decided to remove these. I felt it was repetitive. After all forks are updated, I'm planning to follow up with consistency fixes for previous forks. Not wanting to touch previous specs in this PR.

        if not verify_partial_data_column_header_inclusion_proof(header):
            raise GossipReject("invalid header inclusion proof")

        # [REJECT] The header is proposed by the expected proposer_index
        # (if shuffling is not available, IGNORE instead and MAY be queued for later)
        parent_state = store.block_states[block_header.parent_root].copy()
        process_slots(parent_state, block_header.slot)
        expected_proposer = get_beacon_proposer_index(parent_state)
        if block_header.proposer_index != expected_proposer:
            raise GossipReject("header proposer_index does not match expected proposer")

>         # Cache the validated header for future partial messages

Reviewer: ```suggestion
        # Mark this header as seen
```
This follows this convention but not a strong opinion. Just flagging for consistency.

        seen.partial_data_column_headers[block_root] = header

    if has_cells:
        # [IGNORE] A valid corresponding PartialDataColumnHeader has been seen
        header = seen.partial_data_column_headers.get(block_root)
        if header is None:

>             raise GossipIgnore("no validated header seen for this block")

Reviewer: ```suggestion
            raise GossipIgnore("no valid corresponding header seen for this block")
```
Just a suggestion.

Author: This does need improvements. I think I want to go with:

"valid corresponding header has not been seen"

This matches:

<img width="554" height="110" alt="image" src="https://github.com/user-attachments/assets/1c88cecf-4b88-4893-9532-77dd1340bb83" />

And others.

        block_header = header.signed_block_header.message

        # [IGNORE] The corresponding header is not from a future slot
        # (MAY be queued for processing at the appropriate slot)
        if not is_not_from_future_slot(state, block_header.slot, current_time_ms):

>             raise GossipIgnore("cached header is from a future slot")

Reviewer: ```suggestion
            raise GossipIgnore("the corresponding header is from a future slot")
```
Just a suggestion aligned with the [IGNORE] comment right above.

        # [IGNORE] The corresponding header is from a slot greater than the latest finalized slot
        finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
        if block_header.slot <= finalized_slot:
            raise GossipIgnore(

>                 "cached header is not from a slot greater than the latest finalized slot"

Reviewer: ```suggestion
                "the corresponding header is not from a slot greater than the latest finalized slot"
```

            )

        # [REJECT] The cells_present_bitmap length equals the number of header kzg_commitments
        if len(sidecar.cells_present_bitmap) != len(header.kzg_commitments):

>             raise GossipReject("bitmap length does not match number of commitments")

Reviewer: ```suggestion
            raise GossipReject("bitmap length does not match the number of commitments")
```
I guess?

Author: Yes, that's better. But this is still a little awkward. I'm going to do:

"bitmap length does not match commitments length"

>         # [REJECT] The sidecar's cell and proof data is valid

Reviewer: ```suggestion
        # [REJECT] The sidecar is valid as verified by verify_partial_data_column_sidecar_kzg_proofs
```
This matches the `verify_data_column_sidecar_kzg_proofs` validation in Fulu.

Author: Going to keep this as-is. Including the function name in the comment is repetitive.
